
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
        <meta name="author" content="CIDR">
      
      
        <link rel="canonical" href="https://gstt-cidr.github.io/analysis_guides/virus_mapping/">
      
      
        <link rel="prev" href="..">
      
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.5.3, mkdocs-material-9.5.4">
    
    
      
        <title>Virus mapping - CIDR analysis and SOP hub</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.50c56a3b.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
   <link href="../assets/stylesheets/glightbox.min.css" rel="stylesheet"/><style>
    html.glightbox-open { overflow: initial; height: 100%; }
    .gslide-title { margin-top: 0px; user-select: text; }
    .gslide-desc { color: #666; user-select: text; }
    .gslide-image img { background: white; }
    .gscrollbar-fixer { padding-right: 15px; }
    .gdesc-inner { font-size: 0.75rem; }
    body[data-md-color-scheme="slate"] .gdesc-inner { background: var(--md-default-bg-color);}
    body[data-md-color-scheme="slate"] .gslide-title { color: var(--md-default-fg-color);}
    body[data-md-color-scheme="slate"] .gslide-desc { color: var(--md-default-fg-color);}</style> <script src="../assets/javascripts/glightbox.min.js"></script></head>
  
  
    <body dir="ltr">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#virus-mapping-guide" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="CIDR analysis and SOP hub" class="md-header__button md-logo" aria-label="CIDR analysis and SOP hub" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            CIDR analysis and SOP hub
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Virus mapping
            
          </span>
        </div>
      </div>
    </div>
    
    
      <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
    
      <div class="md-header__source">
        <a href="https://github.com/GSTT-CIDR/analysis_guides.git" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="CIDR analysis and SOP hub" class="md-nav__button md-logo" aria-label="CIDR analysis and SOP hub" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    CIDR analysis and SOP hub
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/GSTT-CIDR/analysis_guides.git" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Introduction
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Virus mapping
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Virus mapping
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#a-guide-for-producing-alignments-from-nanopore-fastq-data" class="md-nav__link">
    <span class="md-ellipsis">
      A guide for producing alignments from nanopore FASTQ data
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pre-requisites" class="md-nav__link">
    <span class="md-ellipsis">
      Pre-requisites
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#step-1-mapping-with-minimap2" class="md-nav__link">
    <span class="md-ellipsis">
      Step 1 - Mapping with Minimap2
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#step-2-alignment-coverage-map" class="md-nav__link">
    <span class="md-ellipsis">
      Step 2 - Alignment coverage map
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#step-4-variant-calling" class="md-nav__link">
    <span class="md-ellipsis">
      Step 4 - Variant calling
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Step 4 - Variant calling">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#freebayes-flags-explanation" class="md-nav__link">
    <span class="md-ellipsis">
      FreeBayes Flags Explanation
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#step-5-generating-a-consensus-sequence" class="md-nav__link">
    <span class="md-ellipsis">
      Step 5 - Generating a consensus sequence
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#step-5-snp-annotation" class="md-nav__link">
    <span class="md-ellipsis">
      Step 5 - SNP annotation
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#a-guide-for-producing-alignments-from-nanopore-fastq-data" class="md-nav__link">
    <span class="md-ellipsis">
      A guide for producing alignments from nanopore FASTQ data
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pre-requisites" class="md-nav__link">
    <span class="md-ellipsis">
      Pre-requisites
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#step-1-mapping-with-minimap2" class="md-nav__link">
    <span class="md-ellipsis">
      Step 1 - Mapping with Minimap2
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#step-2-alignment-coverage-map" class="md-nav__link">
    <span class="md-ellipsis">
      Step 2 - Alignment coverage map
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#step-4-variant-calling" class="md-nav__link">
    <span class="md-ellipsis">
      Step 4 - Variant calling
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Step 4 - Variant calling">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#freebayes-flags-explanation" class="md-nav__link">
    <span class="md-ellipsis">
      FreeBayes Flags Explanation
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#step-5-generating-a-consensus-sequence" class="md-nav__link">
    <span class="md-ellipsis">
      Step 5 - Generating a consensus sequence
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#step-5-snp-annotation" class="md-nav__link">
    <span class="md-ellipsis">
      Step 5 - SNP annotation
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


<!-- HTML imports for lightbox image display -->
<p><head>
<link href="assets/stylesheets/glightbox.min.css" rel="stylesheet"/><style></p>
<script src="assets/javascripts/glightbox.min.js"></script>
<p> <link href="../assets/stylesheets/glightbox.min.css" rel="stylesheet"/><style>
    html.glightbox-open { overflow: initial; height: 100%; }
    .gslide-title { margin-top: 0px; user-select: text; }
    .gslide-desc { color: #666; user-select: text; }
    .gslide-image img { background: white; }
    .gscrollbar-fixer { padding-right: 15px; }
    .gdesc-inner { font-size: 0.75rem; }
    body[data-md-color-scheme="slate"] .gdesc-inner { background: var(--md-default-bg-color);}
    body[data-md-color-scheme="slate"] .gslide-title { color: var(--md-default-fg-color);}
    body[data-md-color-scheme="slate"] .gslide-desc { color: var(--md-default-fg-color);}</style> <script src="../assets/javascripts/glightbox.min.js"></script></head></p>
<h1 id="virus-mapping-guide">Virus mapping guide</h1>
<h2 id="a-guide-for-producing-alignments-from-nanopore-fastq-data">A guide for producing alignments from nanopore FASTQ data</h2>
<p><strong>You will:</strong></p>
<ol>
<li>Identify a virus.</li>
<li>Map against NCBI reference.</li>
<li>Preform variant calling and annotation.</li>
<li>Plot coverage, SNP frequencies and gene annotations.</li>
</ol>
<h2 id="pre-requisites">Pre-requisites</h2>
<ol>
<li>
<p><strong>Mamba/Conda:</strong>
See instruction for <a href="https://mamba.readthedocs.io/en/latest/installation/mamba-installation.html">installing Mamba</a>.</p>
</li>
<li>
<p><strong>Setting up conda environment:</strong></p>
</li>
<li>
<p>Downloading a reference sequence from <a href="https://www.ncbi.nlm.nih.gov/refseq/">RefSeq</a>.</p>
</li>
</ol>
<h2 id="step-1-mapping-with-minimap2">Step 1 - Mapping with Minimap2</h2>
<div class="language-text highlight"><pre><span></span><code><span id="__span-0-1"><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a>minimap2 -ax map-ont ~/refseq/virus/virus_reference.fasta  sequencing_reads.fastq | \
</span><span id="__span-0-2"><a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>samtools view -S -b - |\
</span><span id="__span-0-3"><a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>samtools sort - -o output_alignment.bam 
</span></code></pre></div>
<h2 id="step-2-alignment-coverage-map">Step 2 - Alignment coverage map</h2>
<p>Run BEDtools to generate a coverage map:
<div class="language-text highlight"><pre><span></span><code><span id="__span-1-1"><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a>bedtools genomecov -ibam output_alignment.bam  -bga &gt; genome_coverage.bed
</span></code></pre></div>
Quick-check the coverage stats for your alignment:</p>
<h2 id="step-4-variant-calling">Step 4 - Variant calling</h2>
<p>Perform variant calling using freebayes. Add flags where appropriate:</p>
<h3 id="freebayes-flags-explanation">FreeBayes Flags Explanation</h3>
<p><code>-C</code> or <code>--min-alternate-count</code>
- <strong>Description</strong>: Sets the minimum number of observations that must support an alternate allele within a single individual to consider it in variant calling. This parameter helps to reduce false positives, especially important in noisy sequencing data.
- <strong>Typical Usage</strong>: <code>-C 10</code> implies that at least 10 reads must support an alternate allele before it is considered a valid variant. This threshold can be adjusted based on sequencing quality and expected coverage.</p>
<p><code>-q</code> or <code>--min-mapping-quality</code>
- <strong>Description</strong>: Specifies the minimum mapping quality of reads that are considered in the variant calling process. This flag helps filter out reads with poor alignment scores, which might indicate ambiguous alignments or sequencing artifacts.
- <strong>Typical Usage</strong>: <code>-q 20</code> means only reads with a mapping quality score of 20 or higher will be considered, helping ensure that only high-quality data contributes to the variant calling results.</p>
<p><code>-R</code> or <code>--min-alternate-fraction</code>
- <strong>Description</strong>: Sets the minimum fraction of observations that must support an alternate allele for it to be considered a variant. This setting helps avoid calling very rare variants that might be due to sequencing errors.
- <strong>Typical Usage</strong>: <code>-R 0.1</code> indicates that at least 10% of all reads at a position must support the alternate allele to count it as a variant.</p>
<p><code>--min-coverage</code>
- <strong>Description</strong>: Establishes the minimum number of reads covering a position required to evaluate it for variant calling. Ensuring sufficient coverage at a site is crucial for making reliable variant calls.
- <strong>Typical Usage</strong>: <code>--min-coverage 20</code> means that a site must be covered by at least 20 reads to be considered for variant calling, which helps reduce the likelihood of false positives due to low coverage.</p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-2-1"><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a>freebayes -f [reference.fasta] output_alignment.bam &gt; variants.vcf
</span></code></pre></div>
<div class="language-text highlight"><pre><span></span><code><span id="__span-3-1"><a id="__codelineno-3-1" name="__codelineno-3-1" href="#__codelineno-3-1"></a># Call variants
</span><span id="__span-3-2"><a id="__codelineno-3-2" name="__codelineno-3-2" href="#__codelineno-3-2"></a>bcftools mpileup -Ou -f [reference.fasta] sorted_mapped.bam | bcftools call -mv -Ob | bcftools view -Oz -o variants.vcf.gz
</span></code></pre></div>
<h2 id="step-5-generating-a-consensus-sequence">Step 5 - Generating a consensus sequence</h2>
<div class="language-text highlight"><pre><span></span><code><span id="__span-4-1"><a id="__codelineno-4-1" name="__codelineno-4-1" href="#__codelineno-4-1"></a># First, index the VCF file
</span><span id="__span-4-2"><a id="__codelineno-4-2" name="__codelineno-4-2" href="#__codelineno-4-2"></a>bcftools index variants.vcf.gz
</span><span id="__span-4-3"><a id="__codelineno-4-3" name="__codelineno-4-3" href="#__codelineno-4-3"></a>
</span><span id="__span-4-4"><a id="__codelineno-4-4" name="__codelineno-4-4" href="#__codelineno-4-4"></a># Generate consensus
</span><span id="__span-4-5"><a id="__codelineno-4-5" name="__codelineno-4-5" href="#__codelineno-4-5"></a>cat [reference.fasta] | bcftools consensus -f - variants.vcf.gz &gt; consensus.fa
</span><span id="__span-4-6"><a id="__codelineno-4-6" name="__codelineno-4-6" href="#__codelineno-4-6"></a>
</span><span id="__span-4-7"><a id="__codelineno-4-7" name="__codelineno-4-7" href="#__codelineno-4-7"></a># Generate a bed file where coverage is below the threshold
</span><span id="__span-4-8"><a id="__codelineno-4-8" name="__codelineno-4-8" href="#__codelineno-4-8"></a>awk -v min_depth=[min_depth] &#39;$4 &lt; min_depth {print $1&quot;\t&quot;$2&quot;\t&quot;$3}&#39; genome_coverage.bed &gt; low_coverage.bed
</span><span id="__span-4-9"><a id="__codelineno-4-9" name="__codelineno-4-9" href="#__codelineno-4-9"></a>
</span><span id="__span-4-10"><a id="__codelineno-4-10" name="__codelineno-4-10" href="#__codelineno-4-10"></a># Mask the consensus sequence
</span><span id="__span-4-11"><a id="__codelineno-4-11" name="__codelineno-4-11" href="#__codelineno-4-11"></a>bedtools maskfasta -fi consensus.fa -bed low_coverage.bed -fo masked_consensus.fa
</span></code></pre></div>
<h2 id="step-5-snp-annotation">Step 5 - SNP annotation</h2>
<div class="language-text highlight"><pre><span></span><code><span id="__span-5-1"><a id="__codelineno-5-1" name="__codelineno-5-1" href="#__codelineno-5-1"></a># Download the appropriate database
</span><span id="__span-5-2"><a id="__codelineno-5-2" name="__codelineno-5-2" href="#__codelineno-5-2"></a>snpEff download [genome_version]
</span><span id="__span-5-3"><a id="__codelineno-5-3" name="__codelineno-5-3" href="#__codelineno-5-3"></a>
</span><span id="__span-5-4"><a id="__codelineno-5-4" name="__codelineno-5-4" href="#__codelineno-5-4"></a># Annotate the VCF file
</span><span id="__span-5-5"><a id="__codelineno-5-5" name="__codelineno-5-5" href="#__codelineno-5-5"></a>java -Xmx4g -jar snpEff.jar [genome_version] variants.vcf &gt; annotated_variants.vcf
</span></code></pre></div>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "..", "features": [], "search": "../assets/javascripts/workers/search.c011b7c0.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../assets/javascripts/bundle.7389ff0e.min.js"></script>
      
    
  <script>document$.subscribe(() => {const lightbox = GLightbox({"touchNavigation": true, "loop": false, "zoomable": true, "draggable": true, "openEffect": "zoom", "closeEffect": "zoom", "slideEffect": "slide"});})</script></body>
</html>